### PR TITLE
Implement Kilkari Inbox -- fixes #177, fixes #178, fixes #179, fixes #157

### LINKED ISSUE #157: NMS.KK.ACCESS.004 - Subscription start date following MCTS upload
If the subscriber does NOT call the IVR long code/s, and is subscribed automatically from the back end via an upload of data from MCTS, then the subscription pack for a beneficiary and the week that the message will start in will be based on the LMP or DOB provided in MCTS Data upon data upload (refer #110 NMS.GEN.MCTS.001) as per following rules:
- If DOB is present (with or without LMP) in the data, then Child pack (12-Months-Pack) shall be subscribed. Service shall be “activated” from the date of data upload (subscription state = Active).
- If LMP is present without DOB then pregnancy pack shall be subscribed. Service shall be “activated” from the date of data upload (subscription state = Active) except for case of “early subscription” (for which refer #160 NMS.KK.ACCESS.007).

If a woman is subscribed to Kilkari via MCTS data, the service could start at any point during the 72 week or 48 week schedule. Since we cannot assume that her ASHA or ANM will have informed her about Kilkari, we will need to play her an initial welcome message before any other content is played. This welcome message will introduce her to Kilkari. Once this welcome message has been played in the first week of the service starting, then the next week’s content, as per the LMP or DOB, will be played.

### LINKED ISSUE #177: NMS.KK.IVR.006 - Completed user calling inbox
In case a beneficiary whose subscription has “completed” calls the inbox, Inbox will contain the last message played to user till 7 days after completion. After 7 days the inbox will be deleted and promotional tune shall be played.

### LINKED ISSUE #178: NMS.KK.INBOX.001 - Inbox message rule
Kilkari shall make available a single message of current week in the inbox corresponding to each pack, as soon it is scheduled for OBD delivery.

Even if a pack is configured to deliver multiple OBD messages per week, only the last scheduled message is saved in the inbox.

There shall be no user prompt/welcome message in the inbox.

### LINKED ISSUE #179: NMS.KK.INBOX.003 - Which messages are placed in the inbox
Kilkari shall play the inbox messages on receiving incoming call on Inbox long code

Messages corresponding to multiple packs for a MSISDN shall be played one after the other when Inbox is accessed.

If one subscription is case of early subscription, then that pack shall have no message in the inbox.

### PR DESCRIPTION
**Pull request by llavoie**
Also addresses a few small checkstyle issues here and there

### CODE REVIEW

File: api/src/main/java/org/motechproject/nms/api/web/KilkariController.java [hunk 1]
```diff
@@ -74,14 +80,27 @@ public InboxResponse getInboxDetails(@RequestParam(required = false) Long callin
             throw new NotFoundException(String.format(NOT_FOUND, "callingNumber"));
         }
 
-        Set<Subscription> subscriptions = subscriber.getActiveSubscriptions();
+        Set<Subscription> subscriptions = subscriber.getAllSubscriptions();
```
Comment (larubbio) [inline api/src/main/java/org/motechproject/nms/api/web/KilkariController.java]: What happens with a closed subscription?  Does the try at line 91 handle that?

**Comment by koshalt** [comment]
ok test this

**Comment by koshalt** [comment]
ok to test

**Comment by llavoie** [comment]
FYI, there's clearly something funky with stale code on Jenkins. My CI job just failed on a test case that has an @Ignore attribute on it.

File: api/src/main/java/org/motechproject/nms/api/web/KilkariController.java (hunk 1 above)
Comment (llavoie) [inline api/src/main/java/org/motechproject/nms/api/web/KilkariController.java]: Yep -- once you read InboxServiceImpl.getInboxMessage you'll see how it works. We need to return the final message in a campaign for 7 days after a subscription has completed. After that we return no inbox.

File: api/src/test/java/org/motechproject/nms/api/osgi/KilkariControllerBundleIT.java [hunk 2]
```diff
@@ -38,10 +39,8 @@
 
 import javax.inject.Inject;
 import java.io.IOException;
-import java.util.Arrays;
-import java.util.HashSet;
-import java.util.List;
-import java.util.Set;
+import java.util.*;
```
Comment (larubbio) [inline api/src/test/java/org/motechproject/nms/api/osgi/KilkariControllerBundleIT.java]: Doesn't PMD complain about .* imports?  There is a setting in intelij to stop it from doing this.

File: kilkari/src/main/java/org/motechproject/nms/kilkari/domain/Subscriber.java [hunk 3]
```diff
@@ -89,6 +89,21 @@ public void setLanguage(Language language) {
         this.language = language;
     }
 
+    @Ignore
+    public Set<Subscription> getAllSubscriptions() {
+        // TODO: I have no idea why I need to do this, but returning just this.subscriptions always results in
```
Comment (larubbio) [inline kilkari/src/main/java/org/motechproject/nms/kilkari/domain/Subscriber.java]: Can you put a ticket in, or ask about this on the mailing list?  You can also try to repro it in motech/platform/mds/mds-test-bundle

File: kilkari/src/main/java/org/motechproject/nms/kilkari/exception/NoInboxForSubscriptionException.java [hunk 4]
```diff
@@ -0,0 +1,16 @@
+package org.motechproject.nms.kilkari.exception;
+
+/**
+ *
```
Comment (larubbio) [inline kilkari/src/main/java/org/motechproject/nms/kilkari/exception/NoInboxForSubscriptionException.java]: All comments should be this concise

File: kilkari/src/main/java/org/motechproject/nms/kilkari/service/impl/InboxServiceImpl.java [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+                subscription.getStatus() == SubscriptionStatus.PENDING_ACTIVATION) {
+            // there is no inbox for this subscription, throw
+            throw new NoInboxForSubscriptionException(String.format("No inbox exists for subscription %s",
+                    subscription.getSubscriptionId()));
+        }
+
+        SubscriptionPack pack = subscription.getSubscriptionPack();
+        int daysIntoPack = Days.daysBetween(subscription.getStartDate(), LocalDate.now()).getDays();
+        int messageIndex;
+        int currentWeek = daysIntoPack / DAYS_IN_WEEK + 1;
```
Comment (larubbio) [inline kilkari/src/main/java/org/motechproject/nms/kilkari/service/impl/InboxServiceImpl.java]: I understand operator precedence, but could you add parens just to make it explicit

**Comment by larubbio** [comment]
lgtm

File: api/src/test/java/org/motechproject/nms/api/osgi/KilkariControllerBundleIT.java (hunk 2 above)
Comment (llavoie) [inline api/src/test/java/org/motechproject/nms/api/osgi/KilkariControllerBundleIT.java]: I think I have that Intellij setting correct on one of my clients but not the other. And I don't think we run PMD on our test code, which is why this was missed. Will fix in my next checkin.

File: kilkari/src/main/java/org/motechproject/nms/kilkari/domain/Subscriber.java (hunk 3 above)
Comment (llavoie) [inline kilkari/src/main/java/org/motechproject/nms/kilkari/domain/Subscriber.java]: Yeah, I'll try to come up with a simple repro and then put in a ticket for it.

File: kilkari/src/main/java/org/motechproject/nms/kilkari/service/impl/InboxServiceImpl.java (hunk 5 above)
Comment (llavoie) [inline kilkari/src/main/java/org/motechproject/nms/kilkari/service/impl/InboxServiceImpl.java]: Ok

**Pull request merged**